File: drestaurant-libs/drestaurant-courier/src/main/java/com/drestaurant/courier/domain/CourierOrder.java
package com.drestaurant.courier.domain;

import static org.axonframework.commandhandling.model.AggregateLifecycle.apply;

import org.apache.commons.lang.builder.EqualsBuilder;
import org.apache.commons.lang.builder.HashCodeBuilder;
import org.apache.commons.lang.builder.ToStringBuilder;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.commandhandling.model.AggregateIdentifier;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.spring.stereotype.Aggregate;

import com.drestaurant.courier.domain.api.AssignCourierOrderToCourierCommand;
import com.drestaurant.courier.domain.api.CourierOrderAssignedEvent;
import com.drestaurant.courier.domain.api.CourierOrderCreatedEvent;
import com.drestaurant.courier.domain.api.CourierOrderDeliveredEvent;
import com.drestaurant.courier.domain.api.CreateCourierOrderCommand;
import com.drestaurant.courier.domain.api.MarkCourierOrderAsDeliveredCommand;
import com.drestaurant.courier.domain.model.CourierOrderState;

@Aggregate
class CourierOrder {

	@AggregateIdentifier
	private String id;
	private String cuourierId;
	private CourierOrderState state;

	public CourierOrder() {
	}

	@CommandHandler
	public CourierOrder(CreateCourierOrderCommand command) {
		apply(new CourierOrderCreatedEvent(command.getTargetAggregateIdentifier(), command.getAuditEntry()));
	}

	@EventSourcingHandler
	public void on(CourierOrderCreatedEvent event) {
		this.id = event.getAggregateIdentifier();
		this.state = CourierOrderState.CREATED;
	}

	@CommandHandler
	public void assignToCourier(AssignCourierOrderToCourierCommand command) {
		if (CourierOrderState.CREATED.equals(state)) {
			apply(new CourierOrderAssigningInitiatedEvent(command.getCourierId(), command.getTargetAggregateIdentifier(), command.getAuditEntry()));
		} else {
			throw new UnsupportedOperationException("The current state is not CREATED");
		}
	}

	@EventSourcingHandler
	public void on(CourierOrderAssigningInitiatedEvent event) {
		this.state = CourierOrderState.ASSIGN_PENDING;
	}

	@CommandHandler
	public void markOrderAsAssigned(MarkCourierOrderAsAssignedCommand command) {
		if (CourierOrderState.ASSIGN_PENDING.equals(state)) {
			apply(new CourierOrderAssignedEvent(command.getTargetAggregateIdentifier(), command.getCourierId(), command.getAuditEntry()));
		} else {
			throw new UnsupportedOperationException("The current state is not ASSIGN_PENDING");
		}
	}

	@EventSourcingHandler
	public void on(CourierOrderAssignedEvent event) {
		this.cuourierId = event.getCourierId();
		this.state = CourierOrderState.ASSIGNED;
	}
	
	@CommandHandler
	public void markOrderAsNotAssigned(MarkCourierOrderAsNotAssignedCommand command) {
		if (CourierOrderState.ASSIGN_PENDING.equals(state)) {
			apply(new CourierOrderNotAssignedEvent(command.getTargetAggregateIdentifier(), command.getAuditEntry()));
		} else {
			throw new UnsupportedOperationException("The current state is not ASSIGN_PENDING");
		}
	}
	
	@EventSourcingHandler
	public void on(CourierOrderNotAssignedEvent event) {
		this.state = CourierOrderState.CREATED;
	}
	
	@CommandHandler
	public void markOrderAsDelivered(MarkCourierOrderAsDeliveredCommand command) {
		if (CourierOrderState.ASSIGNED.equals(state)) {
			apply(new CourierOrderDeliveredEvent(command.getTargetAggregateIdentifier(), command.getAuditEntry()));
		} else {
			throw new UnsupportedOperationException("The current state is not ASSIGNED");
		}
	}

	@EventSourcingHandler
	public void on(CourierOrderDeliveredEvent event) {
		this.state = CourierOrderState.DELIVERED;
	}

	@Override
	public String toString() {
		return ToStringBuilder.reflectionToString(this);
	}

	@Override
	public boolean equals(Object o) {
		return EqualsBuilder.reflectionEquals(this, o);
	}

	@Override
	public int hashCode() {
		return HashCodeBuilder.reflectionHashCode(this);
	}

}

File: drestaurant-libs/drestaurant-courier/src/main/kotlin/com/drestaurant/courier/domain/CourierOrder.kt
package com.drestaurant.courier.domain

import com.drestaurant.courier.domain.api.*
import com.drestaurant.courier.domain.model.CourierOrderState
import org.apache.commons.lang.builder.EqualsBuilder
import org.apache.commons.lang.builder.HashCodeBuilder
import org.apache.commons.lang.builder.ToStringBuilder
import org.axonframework.commandhandling.CommandHandler
import org.axonframework.eventsourcing.EventSourcingHandler
import org.axonframework.modelling.command.AggregateIdentifier
import org.axonframework.modelling.command.AggregateLifecycle.apply
import org.axonframework.spring.stereotype.Aggregate

@Aggregate(snapshotTriggerDefinition = "courierOrderSnapshotTriggerDefinition")
internal class CourierOrder {

    @AggregateIdentifier
    private lateinit var id: String
    private lateinit var cuourierId: String
    private lateinit var state: CourierOrderState

    constructor()

    @CommandHandler
    constructor(command: CreateCourierOrderCommand) {
        apply(CourierOrderCreatedEvent(command.targetAggregateIdentifier, command.auditEntry))
    }

    @EventSourcingHandler
    fun on(event: CourierOrderCreatedEvent) {
        id = event.aggregateIdentifier
        state = CourierOrderState.CREATED
    }

    @CommandHandler
    fun assignToCourier(command: AssignCourierOrderToCourierCommand) {
        if (CourierOrderState.CREATED == state) {
            apply(CourierOrderAssigningInitiatedInternalEvent(command.courierId, command.targetAggregateIdentifier, command.auditEntry))
        } else {
            throw UnsupportedOperationException("The current state is not CREATED")
        }
    }

    @EventSourcingHandler
    fun on(event: CourierOrderAssigningInitiatedInternalEvent) {
        state = CourierOrderState.ASSIGN_PENDING
    }

    @CommandHandler
    fun markOrderAsAssigned(command: MarkCourierOrderAsAssignedInternalCommand) {
        if (CourierOrderState.ASSIGN_PENDING == state) {
            apply(CourierOrderAssignedEvent(command.targetAggregateIdentifier, command.courierId, command.auditEntry))
        } else {
            throw UnsupportedOperationException("The current state is not ASSIGN_PENDING")
        }
    }

    @EventSourcingHandler
    fun on(event: CourierOrderAssignedEvent) {
        cuourierId = event.courierId
        state = CourierOrderState.ASSIGNED
    }

    @CommandHandler
    fun markOrderAsNotAssigned(command: MarkCourierOrderAsNotAssignedInternalCommand) {
        if (CourierOrderState.ASSIGN_PENDING == state) {
            apply(CourierOrderNotAssignedEvent(command.targetAggregateIdentifier, command.auditEntry))
        } else {
            throw UnsupportedOperationException("The current state is not ASSIGN_PENDING")
        }
    }

    @EventSourcingHandler
    fun on(event: CourierOrderNotAssignedEvent) {
        state = CourierOrderState.CREATED
    }

    @CommandHandler
    fun markOrderAsDelivered(command: MarkCourierOrderAsDeliveredCommand) {
        if (CourierOrderState.ASSIGNED == state) {
            apply(CourierOrderDeliveredEvent(command.targetAggregateIdentifier, command.auditEntry))
        } else {
            throw UnsupportedOperationException("The current state is not ASSIGNED")
        }
    }

    @EventSourcingHandler
    fun on(event: CourierOrderDeliveredEvent) {
        state = CourierOrderState.DELIVERED
    }

    override fun toString(): String = ToStringBuilder.reflectionToString(this)

    override fun equals(other: Any?): Boolean = EqualsBuilder.reflectionEquals(this, other)

    override fun hashCode(): Int = HashCodeBuilder.reflectionHashCode(this)
}
